### PR TITLE
Lower Handshake to FIRRTL

### PR DESCRIPTION
**Pull request by hanchenye**
Add a lowering pass from Handshake to FIRRTL.

This pass can convert most Handshake elastic operations (except for Load, Store, Memory, and EagerFork) and some Standard binary operations (those supported by a FIRRTL binary primitive) to FIRRTL IR. Currently this pass generates pure combinational circuit, thus in the case when graph cycle exists, the generated FIRRTL IR may have incorrect functionality. Buffer, EagerFork, and other timing components will be added soon.

Implementations of all Handshake elastic operations are written manually. When an elastic operation appears, a corresponding sub-module will be created. Then, this sub-module will be instantiated and inserted into the top-module for substituting the original elastic operation.

Some references: Lana's [dynamically scheduled HLS paper](https://www.epfl.ch/labs/lap/wp-content/uploads/2018/11/JosipovicFeb18_DynamicallyScheduledHighLevelSynthesis_FPGA18.pdf) and her [FPGA20 tutorial paper](https://dynamatic.epfl.ch/papers/FPGA_20_Dynamatic_From_C_C_to_Dynamically_Scheduled_Circuits.pdf).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 build
+.vscode
```
Comment (lattner) [inline .gitignore:2]: It is totally reasonable to drop the .vscode changes, but please split this out to a separate patch (consider it preapproved!)

File: include/circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
+//===- HandshakeToFIRRTL.h --------------------------------------*- C++ -*-===//
```
Comment (lattner) [inline include/circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h:1]: This seems likely to be the only header in this directory.  You might consider moving it up a level to: `include/circt/Conversion/HandshakeToFIRRTL.h`

File: include/circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_CONVERSION_HANDSHAKETOFIRRTL_H_
+#define MLIR_CONVERSION_HANDSHAKETOFIRRTL_H_
+
+#include "circt/Dialect/FIRRTL/Ops.h"
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include <memory>
```
Comment (lattner) [inline include/circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h:24]: None of these #includes are necessary, please drop them.

File: include/circt/Dialect/FIRRTL/OpStructure.td [hunk 4]
```diff
@@ -38,7 +38,7 @@ def CircuitOp : FIRRTLOp<"circuit",
 }
 
 def FModuleOp : FIRRTLOp<"module",
-      [IsolatedFromAbove, FunctionLike, Symbol,
+      [IsolatedFromAbove, FunctionLike, Symbol, RegionKindInterface,
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/OpStructure.td:41]: Please split this out to its own patch, thanks!

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+
+#include "circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h"
+#include "circt/Dialect/FIRRTL/Ops.h"
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
+#include "mlir/IR/Value.h"
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:26]: Please drop transitively included or unnecessary #includes.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 6]
```diff
@@ -0,0 +1,909 @@
...
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/Utils.h"
+
+using namespace mlir;
+using namespace mlir::handshake;
+using namespace circt;
+using namespace circt::firrtl;
+
+using ValueVector = std::vector<Value>;
+using ValueVectorList = std::vector<ValueVector>;
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:38]: Not important for this patch to land but FYI `vector<vector<pointer>>` is very inefficient.  You might consider something like `vector<SmallVector<pointer, 4>>`.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 7]
```diff
@@ -0,0 +1,909 @@
...
+
+//===----------------------------------------------------------------------===//
+// Utils
+//===----------------------------------------------------------------------===//
+
+/// Build a FIRRTL bundle type (with data, valid, and ready subfields) given the
+/// type of the data subfield.
+FIRRTLType buildBundleType(FIRRTLType dataType, bool isFlip,
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:46]: Plz mark methods like this static

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 8]
```diff
@@ -0,0 +1,909 @@
...
+      elements.push_back(BundleElement(dataId, dataType));
+  }
+
+  return BundleType::get(ArrayRef<BundleElement>(elements), context);
+}
+
+/// Return a FIRRTL bundle type (with data, valid, and ready subfields) given a
+/// standard data type. Current supported data types are integer (signed,
+/// unsigned, and signless), index, and none.
+FIRRTLType getBundleType(Type type, bool isFlip) {
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:78]: Random suggestion: this method and buildBundleType would be more ergonomic if you pass in. the builder from the client.  Both ways are fine though.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+  // If the input is already converted to a bundle type elsewhere, itself will
+  // be returned after cast.
+  if (type.isa<BundleType>())
+    return type.cast<BundleType>();
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:82]: plz use `dyn_cast`

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 10]
```diff
@@ -0,0 +1,909 @@
...
+    assert(false && "Unsupported data type. Supported data types: integer "
+                    "(signed, unsigned, signless), index, none");
+    return FIRRTLType(nullptr);
+  }
+}
+
+Value createConstantOp(Location insertLoc, FIRRTLType opType, int value,
+                       ConversionPatternRewriter &rewriter) {
+  IntegerAttr constantOpAttr = rewriter.getIntegerAttr(
+      IntegerType::get(opType.getBitWidthOrSentinel(), rewriter.getContext()),
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:118]: Instead of `IntegerType::get`, I think there is a `rewriter.getIntegerType` method is less syntactically noisy.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 11]
```diff
@@ -542,9 +542,7 @@ void addBranchOps(handshake::FuncOp f, ConversionPatternRewriter &rewriter) {
 
         if (auto condBranchOp = dyn_cast<mlir::CondBranchOp>(termOp))
           newOp = rewriter.create<handshake::ConditionalBranchOp>(
-              termOp->getLoc(),
-              condBranchOp.getCondition(),
-              val);
+              termOp->getLoc(), condBranchOp.getCondition(), val);
```
Comment (lattner) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:545]: This looks like a clang format patch, it would be great to split this off to its own patch.  clang-format-only patches should always be preapproved for CIRCT IMO.

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 12]
```diff
@@ -777,10 +769,8 @@ void checkMergePredecessors(Operation *op) {
 void checkDataflowConversion(handshake::FuncOp f) {
   for (Block &block : f) {
     for (Operation &op : block) {
-      if (!dyn_cast<mlir::CondBranchOp>(op) &&
-          !dyn_cast<mlir::BranchOp>(op) &&
-          !dyn_cast<mlir::LoadOp>(op) &&
-          !dyn_cast<mlir::ConstantOp>(op)) {
+      if (!dyn_cast<mlir::CondBranchOp>(op) && !dyn_cast<mlir::BranchOp>(op) &&
```
Comment (lattner) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:772]: Not your fault, but these should be isa instead of dyn_cast

File: lib/Conversion/StandardToHandshake/StandardToHandshake.cpp [hunk 13]
```diff
@@ -1054,7 +1043,7 @@ void setJoinControlInputs(vector<Operation *> memOps, Operation *memOp,
     auto *op = memOps[i];
     Value val = getBlockControlValue(op->getBlock());
     auto srcOp = val.getDefiningOp();
-    if( !dyn_cast<JoinOp>(srcOp)) {
+    if (!dyn_cast<JoinOp>(srcOp)) {
```
Comment (lattner) [inline lib/Conversion/StandardToHandshake/StandardToHandshake.cpp:1046]: here too, and above.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 14]
```diff
@@ -0,0 +1,909 @@
...
+    }
+  }
+  return FModuleOp(nullptr);
+}
+
+/// All standard expressions and handshake elastic components will be converted
+/// to a FIRRTL sub-module and be instantiated in the top-module.
+FModuleOp createSubModuleOp(FModuleOp topModuleOp, Operation &oldOp,
+                            ConversionPatternRewriter &rewriter) {
+  rewriter.setInsertionPoint(topModuleOp);
```
Comment (stephenneuendorffer) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:229]: Perhaps this should be a custom builder for SubModuleOp that takes lists of types?

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 15]
```diff
@@ -0,0 +1,909 @@
...
+
+        // Check if the sub-module already exists.
+        if (!subModuleOp) {
+          subModuleOp = createSubModuleOp(topModuleOp, op, rewriter);
+
+          Operation *termOp = subModuleOp.getBody().front().getTerminator();
+          Location insertLoc = termOp->getLoc();
+          rewriter.setInsertionPoint(termOp);
+
+          ValueVectorList portList =
```
Comment (stephenneuendorffer) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:803]: It would be nice if the portList could be a data structure which was typesafe when passed to the build...() functions below.

Comment (lattner) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:803]: +1, that could be split out as a separate refactoring in a follow-on patch though.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 14 above)
Comment (hanchenye) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:223]: Yes, this is possible. The current builder of FModuleOp is taking a list of {portName, portType} pairs. Not sure whether this createSubModuleOp() method is common enough for FIRRTL dialect users to become a custom builder of FModuleOp.

Comment (hanchenye) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:223]: This method is especially useful when lowering high-level IRs to FIRRTL IR.

File: include/circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h (hunk 2 above)
Comment (hanchenye) [reply include/circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h:1]: Make sense to me. StandardToHandshake.h is in a similar situation, I can move both of them to `include/circt/Conversion` in another patch.

**Comment by hanchenye** [comment]
All unrelated changes have been moved out.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 16]
```diff
@@ -0,0 +1,893 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.h"
+#include "circt/Dialect/FIRRTL/Ops.h"
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:22]: Pretty sure you don't need Value.h, please minimize #includes.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 17]
```diff
@@ -0,0 +1,893 @@
...
+//===----------------------------------------------------------------------===//
+// Utils
+//===----------------------------------------------------------------------===//
+
+/// Build a FIRRTL bundle type (with data, valid, and ready subfields) given the
+/// type of the data subfield.
+static FIRRTLType buildBundleType(FIRRTLType dataType, bool isFlip,
+                                  MLIRContext *context) {
+  using BundleElement = std::pair<Identifier, FIRRTLType>;
+  llvm::SmallVector<BundleElement, 3> elements;
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:44]: nit: elements could be a c array of 3 elements instead of a smallvector.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 18]
```diff
@@ -0,0 +1,893 @@
...
+  // Add data subfield to the bundle if dataType is not a null.
+  if (dataType) {
+    Identifier dataId = Identifier::get("data", context);
+    if (isFlip)
+      elements.push_back(BundleElement(dataId, FlipType::get(dataType)));
+    else
+      elements.push_back(BundleElement(dataId, dataType));
+  }
+
+  return BundleType::get(ArrayRef<BundleElement>(elements), context);
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:67]: I don't think the explicit ArrayRef cast is necessary

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 19]
```diff
@@ -0,0 +1,893 @@
...
+  }
+  // Currently we consider index type as 64-bits signed integer.
+  case StandardTypes::Index: {
+    int width = type.cast<IndexType>().kInternalStorageBitWidth;
+    return buildBundleType(SIntType::get(context, width), isFlip, context);
+  }
+  case StandardTypes::None:
+    return buildBundleType(/*dataType=*/nullptr, isFlip, context);
+  default:
+    assert(false && "Unsupported data type. Supported data types: integer "
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:104]: Would it be reasonable to return null and emit errors when this method fails?  That would be nicer experience and use the mlir location tracking functionality vs having the compiler crash on invalid input.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 20]
```diff
@@ -0,0 +1,893 @@
...
+                    "(signed, unsigned, signless), index, none");
+    return FIRRTLType(nullptr);
+  }
+}
+
+static Value createConstantOp(Location insertLoc, FIRRTLType opType, int value,
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:110]: is "int value" really enough here?  What about wider things?

Comment (lattner) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:110]: Specifically, it looks like buildConstantLogic will implicitly narrow an APInt to an int, then this will widen it out, losing data.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 21]
```diff
@@ -0,0 +1,893 @@
...
+      rewriter.getIntegerType(opType.getBitWidthOrSentinel()), value);
+  return rewriter.create<firrtl::ConstantOp>(insertLoc, opType, constantOpAttr);
+}
+
+/// Construct a name for creating FIRRTL sub-module. The returned string
+/// contains the following information: 1) standard or handshake operation
+/// name; 2) number of inputs; 3) number of outputs; 4) comparison operation
+/// type (if applied); 5) whether the elastic component is for the control path
+/// (if applied).
+static std::string getSubModuleName(Operation &oldOp) {
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:122]: nit: It is more typical to pass `Operation *` than `Operation&` around.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 22]
```diff
@@ -0,0 +1,893 @@
...
+
+//===----------------------------------------------------------------------===//
+// FIRRTL Top-module Related Functions
+//===----------------------------------------------------------------------===//
+
+/// Currently we are not considering clock and registers, thus the generated
+/// circuit is pure combinational logic. If graph cycle exists, at least one
+/// buffer is required to be inserted for breaking the cycle, which will be
+/// supported in the next patch.
+FModuleOp createTopModuleOp(handshake::FuncOp funcOp,
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:146]: static?

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 23]
```diff
@@ -0,0 +1,893 @@
...
+        rewriter.getStringAttr("arg" + std::to_string(args_idx));
+    FIRRTLType bundlePortType = getBundleType(portType, /*isFlip=*/true);
+    ports.push_back(ModulePort(portName, bundlePortType));
+    args_idx += 1;
+  }
+
+  // Create a FIRRTL module, and inline the funcOp into it.
+  FModuleOp topModuleOp = rewriter.create<FModuleOp>(
+      funcOp.getLoc(), rewriter.getStringAttr(funcOp.getName()),
+      ArrayRef<ModulePort>(ports));
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:173]: Explicit casts to array ref aren't generally needed.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 24]
```diff
@@ -0,0 +1,893 @@
...
+
+  // Merge the second block (inlined from funcOp) of the top-module into the
+  // entry block.
+  auto blockIterator = topModuleOp.getBody().begin();
+  Block *entryBlock = &*blockIterator;
+  Block *secondBlock = &*(++blockIterator);
+
+  // Replace uses of each argument of the second block with the corresponding
+  // argument of the entry block.
+  for (int i = 0, e = secondBlock->getNumArguments(); i < e; ++i) {
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:185]: nit: I think arguments are in a linked list, so this is unnecessarily N^2.  N is usually small, but can be large (e.g. O(1000)) in real world code.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 25]
```diff
@@ -0,0 +1,893 @@
...
+  return topModuleOp;
+}
+
+//===----------------------------------------------------------------------===//
+// FIRRTL Sub-module Related Functions
+//===----------------------------------------------------------------------===//
+
+/// Check whether a submodule with the same name has been created elsewhere.
+/// Return the matched submodule if true, otherwise return nullptr.
+FModuleOp checkSubModuleOp(FModuleOp topModuleOp, Operation &oldOp) {
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:207]: static?  Here and later too

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 14 above)
Comment (lattner) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:222]: +1, it would be nice to add this as a method on FModuleOp.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 26]
```diff
@@ -0,0 +1,893 @@
...
+        rewriter.getStringAttr("arg" + std::to_string(args_idx));
+    FIRRTLType bundlePortType = getBundleType(portType, /*isFlip=*/true);
+    ports.push_back(ModulePort(portName, bundlePortType));
+    args_idx += 1;
+  }
+
+  return rewriter.create<FModuleOp>(
+      topModuleOp.getLoc(),
+      rewriter.getStringAttr(StringRef(getSubModuleName(oldOp))),
+      ArrayRef<ModulePort>(ports));
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:248]: explicit ArrayRef and StringRef casts aren't needed

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 27]
```diff
@@ -0,0 +1,893 @@
...
+    rewriter.create<ConnectOp>(insertLoc, argReady, resultReady);
+
+    if (i != e - 1)
+      rewriter.setInsertionPointToStart(&whenOp.elseRegion().front());
+  }
+}
+
+/// Assume only one input is active.
+/// Please refer to test_cmerge.mlir test case.
+void buildControlMergeLogic(Operation &oldOp, ValueVectorList portList,
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:307]: Can/Should oldOp be typed as `ControlMergeOp` instead of `Operation&`?

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 28]
```diff
@@ -0,0 +1,893 @@
...
+    AndPrimOp selectReadyOp = rewriter.create<AndPrimOp>(
+        insertLoc, argValid.getType(), argValid, resultReady);
+    rewriter.create<ConnectOp>(insertLoc, selectReady, selectReadyOp);
+    if (i != e - 1)
+      rewriter.setInsertionPointToStart(&branchWhenOp.elseRegion().front());
+  }
+}
+
+/// Please refer to test_branch.mlir test case.
+void buildBranchLogic(Operation &oldOp, ValueVectorList portList,
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:409]: Can/should oldOp be typed as `handshake::BranchOp`?

Comment (lattner) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:409]: Similar question for methods below

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 29]
```diff
@@ -0,0 +1,893 @@
...
+  Value controlReady = controlSubfields[1];
+
+  ValueVector resultSubfields = portList.back();
+  Value resultValid = resultSubfields[0];
+  Value resultReady = resultSubfields[1];
+  Value resultData = resultSubfields[2];
+
+  FIRRTLType constantType =
+      FlipType::get(resultData.getType().cast<FIRRTLType>());
+  int constantAttrValue = oldOp.getAttrOfType<IntegerAttr>("value").getInt();
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:571]: What happens when value is larger than 32-bits?

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 30]
```diff
@@ -0,0 +1,893 @@
...
+    // All ports of the instance operation are flipped.
+    FIRRTLType argType = FlipType::get(arg.getType().cast<BundleType>());
+    elements.push_back(BundleElement(argId, argType));
+    args_idx += 1;
+  }
+
+  // Create a instance operation.
+  FIRRTLType instType =
+      BundleType::get(ArrayRef<BundleElement>(elements), context);
+  firrtl::InstanceOp instOp = rewriter.create<firrtl::InstanceOp>(
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:686]: I'd use `auto instOp` for things like this, but either way is ok.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 31]
```diff
@@ -0,0 +1,893 @@
...
+  int ports_idx = 0;
+  for (auto &element : instType.cast<BundleType>().getElements()) {
+    Identifier elementName = element.first;
+    FIRRTLType elementType = element.second;
+    SubfieldOp subfieldOp = rewriter.create<SubfieldOp>(
+        oldOp.getLoc(), elementType, instOp,
+        rewriter.getStringAttr(elementName.strref()));
+
+    // Connect input ports.
+    if (ports_idx < oldOp.getNumOperands())
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:701]: Plz put braces around this 'if' since the body is multi-line with the comment.

Comment (lattner) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:701]: or move the comment into the else if that is where it goes.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 32]
```diff
@@ -0,0 +1,893 @@
...
+          else if (isa<handshake::MuxOp>(op))
+            buildMuxLogic(portList, insertLoc, rewriter);
+
+          else if (isa<handshake::BranchOp>(op))
+            buildBranchLogic(op, portList, insertLoc, rewriter);
+          else if (isa<handshake::ConditionalBranchOp>(op))
+            buildConditionalBranchLogic(op, portList, insertLoc, rewriter);
+
+          else if (isa<handshake::ForkOp>(op))
+            buildForkLogic(op, portList, insertLoc, rewriter);
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:846]: These should all use 'dyn_cast' and pass a typed operation into the callee instead of passing operation&

Comment (lattner) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:846]: In a followup patch it might be worth investigating whether a visitor for handshake would be a good thing to invest in, similar to the firrtl/rtl ones.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 33]
```diff
@@ -0,0 +1,893 @@
...
+
+          else if (isa<handshake::ConstantOp>(op))
+            buildConstantLogic(op, portList, insertLoc, rewriter);
+          else if (isa<handshake::SinkOp>(op))
+            buildSinkLogic(portList, insertLoc, rewriter);
+
+          else if (isa<handshake::JoinOp>(op))
+            buildJoinLogic(portList, insertLoc, rewriter);
+          else
+            assert(false && "Usupported operation type.");
```
Comment (lattner) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:858]: Can this emit an error instead?

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 14 above)
Comment (hanchenye) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp]: 👌 Will do that.
